### PR TITLE
Improve mips.cs esil ##anal

### PR DESCRIPTION
**Pull request by junchao-loongson**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

- Increase the overflow detection of the add、addi、dadd instruction
- Use the new syntax to implement symbolic expansion  

**test**
- I will add a test for overflow detection

**Closing issues**

...

### CODE REVIEW

File: test/db/esil/mips_32 [hunk 1]
```diff
@@ -1014,16 +1014,12 @@ ar v0=0x00112233
 aeip
 aes
 ar s0
-ar hi
-ar lo
```
Comment (junchao-loongson) [inline test/db/esil/mips_32:1018]: 《MIPS® Architecture For Programmers Volume II-A: The MIPS64® Instruction Set, Revision 3.02》P248
This is how the MUL instruction is described：
The 32-bit word value in GPR rs is multiplied by the 32-bit value in GPR rt, treating both operands as signed values,
to produce a 64-bit result. The least significant 32 bits of the product are sign-extended and written to GPR rd. **The
contents of HI and LO are UNPREDICTABLE after the operation.** No arithmetic exception occurs under any circumstances.
So we should not test the value of HI、LO

**Review by radare — APPROVED** [review]

**Comment by trufae** [comment]
616F6223-8E0A-4730-82A8-51B0DAA337F5

**Comment by junchao-loongson** [comment]
it’s all settled

File: libr/anal/p/anal_mips_cs.c [hunk 2]
```diff
@@ -570,10 +566,12 @@ static int analop_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int len
 		}
 		break;
 	case MIPS_INS_MTLO:
-		r_strbuf_appendf (&op->esil, "%s,lo,=,"ES_SIGN_EXT64 ("lo"), REG (0));
+		r_strbuf_appendf (&op->esil, "%s,lo,=", REG (0));
+		ES_SIGN32_64("lo");
```
Comment (trufae) [inline libr/anal/p/anal_mips_cs.c:570]: Missing space before (

File: test/db/esil/mips_32 (hunk 1 above)
Comment (trufae) [reply test/db/esil/mips_32:1018]: Ok then

File: libr/anal/p/anal_mips_cs.c [hunk 3]
```diff
@@ -396,24 +405,19 @@ static int analop_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int len
...
 		case MIPS_INS_ADD: {
 			PROTECT_ZERO () {
-				r_strbuf_appendf (&op->esil, "%s,%s,+,%s,=",
-					ARG (1), ARG (2), ARG (0));
-#if 0
-			r_strbuf_appendf (&op->esil,
-				"0,32,%s,%s,+,>>,>,?{,1,TRAP,}{,%s,%s,+,%s,=,}",
-				ARG(2), ARG(1), ARG(2), ARG(1), ARG(0));
-#endif
+				ES_ADD_CK_OVERF(ARG(1), ARG(2), ARG(0), 32);
```
Comment (trufae) [inline libr/anal/p/anal_mips_cs.c:408]: Missing space

File: libr/anal/p/anal_mips_cs.c [hunk 4]
```diff
@@ -82,11 +82,20 @@ static ut64 t9_pre = UT64_MAX;
...
-	arg",0x80000000,&,0,<,?{,"\
-		"0xffffffff00000000,"arg",|=,"\
-	"}"
+
+#define ES_SIGN_n_64(arg, n_bit)  do{if (a->bits == 64) \
+		{r_strbuf_appendf (&op->esil, ",%d,%s,~,%s,=,", n_bit, arg, arg);}else{r_strbuf_append(&op->esil,",");}}while(0)
+#define ES_SIGN32_64(arg)	ES_SIGN_n_64(arg, 32)
+#define ES_SIGN16_64(arg)	ES_SIGN_n_64(arg, 16)
+
+#define ES_ADD_CK_OVERF(x, y,z,bit) do{uint64_t mask = 1ULL<<(bit-1);\
```
Comment (trufae) [inline libr/anal/p/anal_mips_cs.c:96]: Use ut64

Comment (trufae) [reply libr/anal/p/anal_mips_cs.c:96]: And use proper spacings

File: libr/anal/p/anal_mips_cs.c [hunk 5]
```diff
@@ -82,11 +82,20 @@ static ut64 t9_pre = UT64_MAX;
...
+#define ES_H(x) "0xffff,"x",&"
+#define ES_W(x) "0xffffffff,"x",&"
+
 // sign extend 32 -> 64
-#define ES_SIGN_EXT64(arg) \
-	arg",0x80000000,&,0,<,?{,"\
-		"0xffffffff00000000,"arg",|=,"\
-	"}"
+
+#define ES_SIGN_n_64(arg, n_bit)  do{if (a->bits == 64) \
```
Comment (trufae) [inline libr/anal/p/anal_mips_cs.c:91]: Make a separate function or properly indent this as multiline

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/anal/p/anal_mips_cs.c (hunk 5 above)
Comment (junchao-loongson) [reply libr/anal/p/anal_mips_cs.c:91]: I want to save a little function call expend

Comment (XVilka) [reply libr/anal/p/anal_mips_cs.c:91]: @junchao-loongson you can just add an inline function instead.

Comment (junchao-loongson) [reply libr/anal/p/anal_mips_cs.c:91]: okay~

**Review by XVilka — APPROVED** [review]

**Pull request merged**